### ISSUE TITLE
The spec for header union operations seems incomplete

### ISSUE THREAD

**Issue by mihaibudiu** [p4lang/p4-spec#1221]
First, it only lists two operations, setValid and setInvalid.
Second, there is no discussion about the semantics of assigning invalid fields. I recall we had a long debate about the meaning of:

```
header_union HU { H1 h1; H2 h2; }
HU u;
H1 h;
h.setInvalid();
u.field =h;
```

The spec does not say whether this invalidates the valid fields of the union.

Notice that the compiler expands a header union into a field-by-field copy: 
```
HU u1; HU u2;
u1 = u2;
```

is translated to 
```
u1.h1 = u2.h1;
u1.h2 = u2.h2;
```

This is correct only if copying invalid headers does not disturb the valid ones.

**Comment by mbudiu-vmw** [comment]
However, the spec implies that this is not true.

**Comment by jafingerhut** [comment]
@mbudiu-vmw wrote: "Second, there is no discussion about the semantics of assigning invalid fields."

It appears to me that the spec if quite explicit on this topic.  Quoting from Section 8.19 "Operations on Header Unions" below:

```
We can understand an assignment to a union field

u.hi = e
as equivalent to

  u.hi.setValid();
  u.hi = e;
if e is valid and

  u.hi.setInvalid();
otherwise.
```
If I am interpreting that correctly, then your example code in the original comment above:
```
header_union HU { H1 h1; H2 h2; }
HU u;
H1 h;
h.setInvalid();
u.field =h;
```
The spec explicitly says that the HU should end in a state where all of its members are invalid.

How can that section of the spec be interpreted otherwise?

**Comment by mbudiu-vmw** [comment]
The spec does not explain whether an existing valid field is invalidated when assigning a different field with an invalid value. Either interpretation would be possible.

**Comment by jafingerhut** [comment]
Please read the entire section on operations on header unions and tell me if you still believe it is open to interpretation.  I don't see how it is.  For example, see this text:

More formally, if u is an expression whose type is a header union U with fields ranged over by hi, then the following operations can be used to manipulate u:

u.hi.setValid(): sets the valid bit for header hi to true and sets the valid bit for all other headers to false, which implies that it is unspecified what value reading any member header of u will return.

u.hi.setInvalid(): if the valid bit for any member header of u is true then sets it to false, which implies that it is unspecified what value reading any member header of u will return.

**Comment by jafingerhut** [comment]
If you mean that this phrase is ambiguous "if the valid bit for any member header of u is true then sets it to false", I believe the intent when it was written was more explicitly the following.

"assign false to the valid bit for every member header of u"

I would be open to changing that sentence to be more explicit if you think it is open to interpretation.

**Comment by mbudiu-vmw** [comment]
Fixed by #1220

**Comment by jafingerhut** [comment]
Closing this issue, as it appears to have been resolved by the commit mentioned by Mihai above.
